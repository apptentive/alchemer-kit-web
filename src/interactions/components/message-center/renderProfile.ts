interface IProfileOptions {
  isRequired: boolean;
  title: string;
  name: {
    placeholder: string;
    value: string;
  };
  email: {
    placeholder: string;
    value: string;
  };
}

/**
 * Renders the Message Center greeting (title, name (label, input), email (label, input)).
 *
 * @param {object} options - The options object to render the profile input
 * @returns {HTMLDivElement} - The profile DOM Node.
 */
const renderProfile = (options: IProfileOptions) => {
  const required = options.isRequired;

  const header = document.createElement('h2');
  header.classList.add('apptentive-message-center-profile__title');
  header.classList.add('apptentive-typography__body');
  header.textContent = options.title;

  const nameInput = document.createElement('input');
  nameInput.type = 'text';
  nameInput.classList.add('apptentive-textinput');
  nameInput.placeholder = options.name.placeholder;

  nameInput.value = options.name.value;

  const nameLabel = document.createElement('label');
  nameLabel.classList.add('apptentive-message-center-profile__name');
  nameLabel.append(nameInput);

  const emailInput = document.createElement('input');
  emailInput.type = 'email';
  emailInput.classList.add('apptentive-textinput');
  emailInput.placeholder = options.email.placeholder;

  emailInput.value = options.email.value;
  emailInput.addEventListener('blur', (event) => {
    (event.target as HTMLInputElement).classList.remove('apptentive-textinput--error');
  });

  if (required) {
    emailInput.required = true;
  }

  const emailLabel = document.createElement('label');
  emailLabel.classList.add('apptentive-message-center-profile__email');
  emailLabel.append(emailInput);

  const profile = document.createElement('div');
  profile.classList.add('apptentive-message-center-profile');
  profile.append(header);
  profile.append(nameLabel);
  profile.append(emailLabel);

  return profile;
};

export default renderProfile;
