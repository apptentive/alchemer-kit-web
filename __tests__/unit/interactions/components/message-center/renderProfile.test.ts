import renderProfile from '../../../../../src/interactions/components/message-center/renderProfile';

const _title = 'Profile Title';
const _namePlaceholder = 'Name placeholder';
const _nameValue = 'Suzy Creamcheese';
const _emailPlaceholder = 'Email placeholder';
const _emailValue = 'suzy.creamcheese@bakery.com';

const _defaultOptions = {
  isRequired: true,
  title: _title,
  name: {
    placeholder: _namePlaceholder,
    value: _nameValue,
  },
  email: {
    placeholder: _emailPlaceholder,
    value: _emailValue,
  },
};

describe('renderProfile', () => {
  test('properly renders profile that is required', () => {
    const output = renderProfile(_defaultOptions);

    expect(output).toMatchSnapshot();
  });

  test('properly renders profile that is optional', () => {
    const output = renderProfile({
      ..._defaultOptions,
      isRequired: false,
    });

    expect(output).toMatchSnapshot();
  });

  test('properly attaches blur event to email field', () => {
    document.body.append(renderProfile(_defaultOptions));

    const emailInput = document.querySelector<HTMLInputElement>('input[type="email"]')!;
    emailInput.classList.add('apptentive-textinput--error');
    emailInput.focus();
    emailInput.blur();

    expect(emailInput.classList.contains('apptentive-textinput--error')).toBe(false);

    document.body.innerHTML = '';
  });
});
