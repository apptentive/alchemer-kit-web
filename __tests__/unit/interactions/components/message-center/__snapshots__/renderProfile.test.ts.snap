// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderProfile properly renders profile that is optional 1`] = `
<div
  class="apptentive-message-center-profile"
>
  <h2
    class="apptentive-message-center-profile__title apptentive-typography__body"
  >
    Profile Title
  </h2>
  <label
    class="apptentive-message-center-profile__name"
  >
    <input
      class="apptentive-textinput"
      placeholder="Name placeholder"
      type="text"
    />
  </label>
  <label
    class="apptentive-message-center-profile__email"
  >
    <input
      class="apptentive-textinput"
      placeholder="Email placeholder"
      type="email"
    />
  </label>
</div>
`;

exports[`renderProfile properly renders profile that is required 1`] = `
<div
  class="apptentive-message-center-profile"
>
  <h2
    class="apptentive-message-center-profile__title apptentive-typography__body"
  >
    Profile Title
  </h2>
  <label
    class="apptentive-message-center-profile__name"
  >
    <input
      class="apptentive-textinput"
      placeholder="Name placeholder"
      type="text"
    />
  </label>
  <label
    class="apptentive-message-center-profile__email"
  >
    <input
      class="apptentive-textinput"
      placeholder="Email placeholder"
      required=""
      type="email"
    />
  </label>
</div>
`;
